### ISSUE TITLE
Multi-Tenancy Support for EF Core

### ISSUE THREAD

**Issue by jeremydmiller** [JasperFx/wolverine#556]
Just trying to create support for addressing either filtered by tenant DbContext values or connecting to separate databases for tenant with EF Core similar to what Wolverine already does with Marten.

**Comment by jeremydmiller** [comment]
## Scenarios:

* Both PostgreSQL & Sql Server for now
* Static tenant id to connection string mapping
* Is it allowed to specify no, or the default tenant id? If so, return the default database? If not, throw an assertion?
* throw UnknownTenantException if the tenant cannot be found
* Mimic Marten's "master table" tenancy model. *Maybe* pull out some common code for that
* Rotating credentials for PostgreSQL? Maybe post 4.0
* Rotating credentials for Sql Server? Definitely post 4.0
* Stupid modular monolith architectures are definitely coming, so be able to use multiple DbContext types for the same storage configuration, so there's some value in being able to register the tenancy separate from the DbContext type and refer to the tenancy

## Ideas

* Establish the Wolverine envelope storage first? Then either chain EF Core registrations off of that, or refer to it by name? 
*  Might have to allow separate modular monoliths to register different DbContext types in non-centralized registrations. You *know* someone clever will want to do that
* Define the tenancy inline w/ db context?
* Let users have a static `Configure(DbContextBuilder<T> builder, string connectionString)` method? Use that for codegen
* Could register a new IDbContextBuilder<T>? Takes in a string for the connection string
* Do we need to split static vs per scope DbContext configuration?

**Comment by jeremydmiller** [comment]
## Test Scenarios

- [ ] More than one DbContext registered for the same set of multi-tenanted databases
- [ ] Register multi-tenancy by database for PostgreSQL
- [ ] Register multi-tenancy by database for Sql Server
- [ ] Working IDatabaseSource for multi-tenanted databases
- [ ] Marten & DbContext running side by side w/ same multi-tenancy?
- [ ] Lightweight saga storage with multi-tenancy
- [ ] Outbox integration w/ multi-tenancy + EF Core + Postgresql w/ Transactional middleware
- [ ] Outbox integration w/ multi-tenancy + EF Core + SQL Server w/ Transactional middleware
- [ ] Usage outside of transactional middleware

**Comment by davidlahuta** [comment]
This feature would be incredibly helpful for my use case!

I have a scenario where there's a "main" PostgreSQL database containing tenant metadata, and each tenant has its own PostgreSQL database.

Currently, I spin up one IHost for "main" messaging and a separate IHost for each tenant's messaging within a single ASP.NET Core application process.

While this works, it comes with some challenges:
- On worker services (where I run the handlers for both "main" and tenant-specific messaging), memory usage is quite high — from 30MB to 200MB  per IHost.
- On API backends (where the IHosts are only used for publishing), the memory usage is minimal — around 2MB per IHost.
- Database server connections aren't pooled across IHosts, so I had to introduce PgBouncer to handle the load efficiently.

Question
Do you have any guidance on how to optimize this setup?
Would it be possible (or advisable) to share multiple Wolverine instances within a single IHost?
Any patterns or recommended approaches to achieve better resource efficiency in this kind of multi-tenant messaging setup before this feature is released?

Thanks in advance!

**Comment by jeremydmiller** [comment]
@davidlahuta This is a much longer conversation, but I don't think there should be a reason to use separate IHosts in the same process for multi-tenancy. Especially after we finally get this story in for 4.0.

**Comment by davidlahuta** [comment]
I agree — once it's available in 4.0, I'll migrate my solution to use this feature.

In the meantime, are you saying it's possible to call AddWolverine(...) multiple times on the same IHost? If so, how can I access the specific Wolverine instance? Let's say I need an instance registered for specific tenant.

**Comment by jeremydmiller** [comment]
## Sample Application

Turns out that because of how embarrassingly pitiful EF Core migrations are, we pretty much have to have a full application in order to do much end to end testing, and we need one for documentation purposes anyway, so we need an app with:

- [ ] Wolverine HTTP too
- [ ] The ItemsDbContext for basic operations against Items
- [ ] A command handler using storage operations
- [ ] An HTTP endpoint using storage operations
- [ ] A command handler using a DbContext directly
- [ ] An HTTP endpoint using a DbContext directly
- [ ] An HTTP GET endpoint -- verify no SaveChangesAsync generated!
- [ ] Using a lightweight saga

**Comment by jeremydmiller** [comment]
This is in. Might want to add some additional tenant admin, but I don't think that makes the cut for 4.0

**Comment by YBazanPro** [comment]
> > * [ ]   Handle one database holding multiple tenants, make sure that the databases aren't really doubled up[ ]   Dynamic addition of tenants at runtime for master table tenancy w/ PostgreSQL[ ]   Dynamic addition of tenants at runtime for master table tenancy w/ Sql Server
> 
> Are you planning this features to be available for 4.0?
> 
> Can you share sample application? I would like to try it out and test it.

Any information on this?

I'm struggling on this very point.

**Comment by jeremydmiller** [comment]
@YBazanPro Hey, don't comment on closed issues. You're lucky I happened to see this come through.

This has been in Wolverine for months now:

https://wolverinefx.net/guide/durability/efcore/multi-tenancy.html
